Meme/Meme/NetworkClient.swift

```diff
@@ -8,49 +8,64 @@
 
 import Foundation
 
-protocol MemeClient {
-    func getMeme(from template: Template) -> Meme?
-}
-
-class NetworkClient: MemeClient {
-
-    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
-        guard let data = data,
-            let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String:String] else {
-                print("Error deserializing template data");
-                return nil
-        }
-        return jsonResponse.compactMap({ return Template.init(name: $0.0, templateURLString: $0.1)})
-    }
+class NetworkClient {
 
-    func requestTemplates(completion: @escaping ([Template]) -> Void)  {
+    func requestTemplates(completion: @escaping (Result<[Template], Error>) -> Void)  {
         let session = URLSession.shared
         let urlBuilder = MemeEndpointBuilder()
         let url = urlBuilder.endpointURL(.template)
         let request = URLRequest(url:url)
         let task: URLSessionDataTask = session.dataTask(with: request) { (data, response, error) -> Void in
-            guard let templates = self.getTemplates(data) else { return }
-            completion(templates)
+            guard let data = data else {
+                if let error = error {
+                    completion(.failure(error))
+                } else {
+                    if let responseError = (response as? HTTPURLResponse)?.apiError {
```

Comment:
The else case is left unhandled. Adding a .unknownError case would allow you to nil-coalesce with that as the default which means no need for an explicit if else block